File: 283-move-zeroes/solution.py
# -*- coding:utf-8; -*-
class Solution:
    def moveZeroes(self, nums):
        preZero = -1  # 指向遍历元素前面最后一个非0元素
        for i, v in enumerate(nums):
            if v != 0:
                preZero += 1
                if preZero != i:  # 避免原地互换
                    nums[preZero], nums[i] = nums[i], nums[preZero]


if __name__ == "__main__":
    nums = [1, 2, 3, 4, 5]
    s = Solution()
    s.moveZeroes(nums)
    print(nums)

File: 283-move-zeroes/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int zeroStart = 0;
        for (int i=0; i < nums.size(); ++i) {
            if (nums[i]!=0){
                if(i!=zeroStart) {
                    iter_swap(nums.begin()+i,nums.begin()+zeroStart);
                }

                zeroStart++;
            }
        }
    }
};
